src/charm.py

```diff
@@ -65,19 +73,67 @@ def _on_website_relation_joined(self, event):
                     'port': str(port)
                 })
 
+    def _on_metrics_endpoint_relation_created(self, event: RelationJoinedEvent):
+        username = metrics_username(event.relation)
+        password = secrets.token_urlsafe(16)
+        self.control_socket.add_metrics_user(username, password)
 
-def api_port():
-    ''' api_port determines the port that the controller's API server is
-        listening on.  If the machine does not appear to be a juju
-        controller then None is returned.
+        # Set up Prometheus scrape config
+        self.metrics_endpoint = MetricsEndpointProvider(self,
+            jobs = [{
+                # "job_name": "juju",
+                "metrics_path": "/introspection/metrics",
+                "scheme": "https",
+                "static_configs": [{"targets": ["*:17070"]}],
+                "basic_auth": {
+                    "username": f'user-{username}',
+                    "password": password,
+                },
+                "tls_config": {
+                    "ca_file": ca_cert(),
+                    "server_name": "juju-apiserver",
+                },
+            }],
+        )
+        self.metrics_endpoint.set_scrape_job_spec()
+        
+    def _on_metrics_endpoint_relation_broken(self, event: RelationDepartedEvent):
+        # Remove metrics user
+        username = metrics_username(event.relation)
+        self.control_socket.remove_metrics_user(username)
+
+
+def _agent_conf(key: str):
     '''
-    machine = os.getenv('JUJU_MACHINE_ID')
-    if machine is None:
-        return None
-    path = '/var/lib/juju/agents/machine-{}/agent.conf'.format(machine)
+    _agent_conf reads a value from the agent.conf file on disk.
+    If the machine does not appear to be a Juju controller, then None is
+    returned.
+    '''
+    # TODO: get the unit number/name from an environment variable?
+    path = '/var/lib/juju/agents/unit-controller-0/agent.conf'
```

Comment:
Maybe a combination of JUJU_MACHINE_ID and JUJU_CHARM_DIR? https://github.com/juju/juju/blob/3.3/worker/uniter/runner/context/context.go#L1295

Reply:
yeah, this is not gonna work in HA

Reply:
Probably want to use one of these hook variables
```
'JUJU_UNIT_NAME': 'controller/0'
'JUJU_AGENT_SOCKET_ADDRESS': '@/var/lib/juju/agents/unit-controller-0/agent.socket'
'JUJU_CHARM_DIR': '/var/lib/juju/agents/unit-controller-0/charm'
'CHARM_DIR': '/var/lib/juju/agents/unit-controller-0/charm'
```

Reply:
actually, we should probably use `self.unit.name` from `ops`.